### PR TITLE
Lwaftr noodling

### PR DESCRIPTION
**Pull request by andywingo**
Before:

statisticsv6: 1.916 MPPS, 9.046 Gbps.
statisticsv4: 1.916 MPPS, 7.819 Gbps.

After:

statisticsv6: 2.277 MPPS, 10.747 Gbps.
statisticsv4: 2.277 MPPS, 9.290 Gbps.

### CODE REVIEW

File: src/apps/lwaftr/lwaftr.lua [hunk 1]
```diff
@@ -19,9 +19,20 @@ local ffi = require("ffi")
...
 local C = ffi.C
+local cast = ffi.cast
+local receive, transmit = link.receive, link.transmit
+local empty, full = link.empty, link.full
 
 local debug = false
 
+local function compute_binding_table_by_ipv4(binding_table)
+   local ret = {}
+   for i,bind in ipairs(binding_table) do
```
Comment (dpino) [inline src/apps/lwaftr/lwaftr.lua]: Nit: variable 'i' is not used in the loop, usually variable '_' is used as a placeholder for ignored variables.

File: src/apps/lwaftr/lwaftr.lua [hunk 2]
```diff
@@ -337,7 +360,7 @@ local function from_inet(lwstate, pkt)
          return -- lookup failed
       else
          local src_ip_start = constants.ethernet_header_size + constants.o_ipv4_src_addr
-         --local to_ip = ffi.cast("uint32_t*", pkt.data + src_ip_start)[0]
+         --local to_ip = cast("uint32_t*", pkt.data + src_ip_start)[0]
```
Comment (dpino) [inline src/apps/lwaftr/lwaftr.lua]: Better to remove this commented line of code.

File: src/apps/lwaftr/lwaftr.lua [hunk 3]
```diff
@@ -554,11 +577,11 @@ function LwAftr:push ()
    self.o4 = self.output.v4
    self.o6 = self.output.v6
 
-   while not link.empty(i4) do --and not link.full(o4) and not link.full(o6) do
-      local pkt = link.receive(i4)
+   while not empty(i4) do --and not full(o4) and not full(o6) do
```
Comment (dpino) [inline src/apps/lwaftr/lwaftr.lua]: If these checkings are not finally used, shouldn't we remove them?

File: src/apps/lwaftr/lwaftr.lua [hunk 4]
```diff
@@ -19,9 +19,20 @@ local ffi = require("ffi")
 local band, bnot, rshift = bit.band, bit.bnot, bit.rshift
 local bitfield = lib.bitfield
 local C = ffi.C
+local cast = ffi.cast
+local receive, transmit = link.receive, link.transmit
+local empty, full = link.empty, link.full
```
Comment (dpino) [inline src/apps/lwaftr/lwaftr.lua]: `link.full` is not finally used, it was previously used in the while loops but commented.

**Comment by kbara** [comment]
From current lwaftr_nutmeg, the dev branch: (07bf01fb8f5c372782803671a1694ace0a50b81f)
```
$ for x in {1..10}; do lwaftr-run; done | lwstats.py
Initial v4 MPPS: min: 1.026, max: 1.284, avg: 1.127, stdev: 0.0865 (n=10)
Initial v4 Gbps: min: 4.184, max: 5.239, avg: 4.600, stdev: 0.3535 (n=10)
Initial v6 MPPS: min: 1.026, max: 1.284, avg: 1.127, stdev: 0.0865 (n=10)
Initial v6 Gbps: min: 4.841, max: 6.061, avg: 5.322, stdev: 0.4089 (n=10)
  Final v4 MPPS: min: 1.055, max: 1.344, avg: 1.164, stdev: 0.0972 (n=10)
  Final v4 Gbps: min: 4.304, max: 5.483, avg: 4.751, stdev: 0.3968 (n=10)
  Final v6 MPPS: min: 1.055, max: 1.344, avg: 1.164, stdev: 0.0972 (n=10)
  Final v6 Gbps: min: 4.979, max: 6.344, avg: 5.496, stdev: 0.4593 (n=10)
```
------------

lwaftr-noodling (without Gbps figures, as those are from a newer lwstats script):
```
$ for x in {1..10}; do lwaftr-run; done | lwstats.py
Initial v4 MPPS: min: 1.421, max: 2.102, avg: 1.657, stdev: 0.2963
Initial v6 MPPS: min: 1.421, max: 1.983, avg: 1.621, stdev: 0.2488
  Final v4 MPPS: min: 1.357, max: 2.177, avg: 1.741, stdev: 0.3718
  Final v6 MPPS: min: 1.357, max: 2.036, avg: 1.689, stdev: 0.3046
```

I'm merging this, and will address some of the nits. Some of the others are remaining as reminders until I've had a chance to analyse their impact on the code running at line speed.

**Comment by kbara** [comment]
Post-merge stats:
```
$ for x in {1..10}; do lwaftr-run; done | lwstats.py
Initial v4 MPPS: min: 1.417, max: 2.134, avg: 1.814, stdev: 0.3341 (n=10)
Initial v4 Gbps: min: 5.780, max: 8.705, avg: 7.399, stdev: 1.3631 (n=10)
Initial v6 MPPS: min: 1.417, max: 2.004, avg: 1.746, stdev: 0.2751 (n=10)
Initial v6 Gbps: min: 6.687, max: 9.458, avg: 8.240, stdev: 1.2981 (n=10)
  Final v4 MPPS: min: 1.494, max: 2.178, avg: 1.905, stdev: 0.3478 (n=10)
  Final v4 Gbps: min: 6.097, max: 8.885, avg: 7.772, stdev: 1.4181 (n=10)
  Final v6 MPPS: min: 1.494, max: 2.036, avg: 1.822, stdev: 0.2764 (n=10)
  Final v6 Gbps: min: 7.053, max: 9.609, avg: 8.599, stdev: 1.3034 (n=10)
```

**Comment by kbara** [comment]
My bench_ui statistics appear to be significantly better than those in the initial post:

On Interlaken:
```
$ sudo ./snabb snsh  apps/lwaftr/benchui.lua ../tests/apps/lwaftr/data/binding.table ../tests/apps/lwaf/data/tunnel_icmp.conf ../tests/apps/lwaftr/benchdata/ipv4-0550.pcap ../tests/apps/lwaftr/benchdata/ipv6-0550.pcap 
statisticsv6: 3.652 MPPS, 17.235 Gbps.
statisticsv4: 3.652 MPPS, 14.899 Gbps.
```

**Comment by kbara** [comment]
Other runs (still Interlaken) are better yet:
```
sudo ./snabb snsh  apps/lwaftr/benchui.lua ../tests/apps/lwaftr/data/binding.table ../tests/apps/lwaf/data/tunnel_icmp.conf ../tests/apps/lwaftr/benchdata/ipv4-0550.pcap ../tests/apps/lwaftr/benchdata/ipv6-0550.pcap 
statisticsv6: 4.133 MPPS, 19.507 Gbps.
statisticsv4: 4.133 MPPS, 16.863 Gbps.
```
Yes, those are initial non-warmed-up statistics.

I'd like to see what would happen with using a non-ipairs loop, and to see a -D (duration) option to benchui.

File: src/apps/lwaftr/lwaftr.lua (hunk 2 above)
Comment (kbara) [inline src/apps/lwaftr/lwaftr.lua]: I've removed it.

File: src/apps/lwaftr/lwaftr.lua (hunk 3 above)
Comment (kbara) [inline src/apps/lwaftr/lwaftr.lua]: Not yet. Analysis should be done when we're hitting line speed.

**Comment by lukego** [comment]
Could be worth upgrading Interlaken to Ubuntu 15.04 for a couple of reasons:

1. To enable `no_turbo` mode. This can be nice for more consistent benchmarking to disable the Xeon TurboBoost feature. TurboBoost dynamically adjusts the CPU frequency based on various factors (e.g. how many cores are busy) and that can make wall-clock based results (e.g. Mpps) inconsistent. I like to disable it but requires a newer kernel afaik. (Alternative probably reasonable solution is to look at e.g. cycles/packet that is probably pretty stable across CPUs.)
2. To upgrade to a newer GLIBC that will have AVX versions of mempcy and memmove. This could improve performance if those happen to be bottlenecks. (Probably we should integrate our own versions into Snabb Switch instead of depending on GLIBC -- I would quite like to write dynasm based ones but it seems nontrivial to add AVX support to dynasm.)

Let me know if you would like such an upgrade and I'll do it.

**Comment by kbara** [comment]
@lukego yes, please, if it won't cause too much downtime. I was just wondering about turbo factors earlier this afternoon. And memcpy/memmove are the single biggest consumers of time in our top two traces. When can you do this, and when would Interlaken be back up?

**Comment by lukego** [comment]
Updating now. ETA half an hour or so.

**Comment by kbara** [comment]
Fantastic, thank you.

**Comment by kbara** [comment]
Did everything go ok with the upgrade? It's taking an awfully long time to come back up.

**Comment by lukego** [comment]
Boot had hung in the BIOS for some reason. Power cycled. Turbo should be disabled: see `/etc/rc.local`.

For memcpy you could also experiment with using the AVX routine in DPDK's `rte_memcpy.h`. That is easy to extract and experiment with (e.g. `C.rte_memcpy()` instead of `ffi.copy()`). I experimented with that in the NFV application and initially saw a significant improvement over GLIBC but I was concerned this would only hold in benchmarks (speedup was with exactly 256-byte packets and seemed to disappear with average 256-byte packets). YMMV. I would like to have our own memcpy routines going forward e.g. based on Agner's asmlib and potentially taking advantage of known alignment of packet data.

**Comment by kbara** [comment]
Thanks for the pointers; I'll check them out after I've got a chunk of documentation written. Performance seems a bit lower with Turbo disabled; it was also lower today before the upgrade, for reasons yet to be determined, but not quite as much lower.
```
$ for x in {1..10}; do lwaftr-run; done | lwstats.py 
Initial v4 MPPS: min: 1.082, max: 1.759, avg: 1.548, stdev: 0.3117 (n=10)
Initial v4 Gbps: min: 4.414, max: 7.176, avg: 6.315, stdev: 1.2719 (n=10)
Initial v6 MPPS: min: 1.082, max: 1.759, avg: 1.548, stdev: 0.3117 (n=10)
Initial v6 Gbps: min: 5.106, max: 8.302, avg: 7.305, stdev: 1.4716 (n=10)
  Final v4 MPPS: min: 1.129, max: 1.859, avg: 1.633, stdev: 0.3411 (n=10)
  Final v4 Gbps: min: 4.608, max: 7.585, avg: 6.662, stdev: 1.3910 (n=10)
  Final v6 MPPS: min: 1.129, max: 1.859, avg: 1.633, stdev: 0.3411 (n=10)
  Final v6 Gbps: min: 5.330, max: 8.775, avg: 7.707, stdev: 1.6091 (n=10)
```
Understanding the performance characteristics of Snabb apps is definitely an interesting process. :-)

Do you mind if I change the performance profile from powersave? it was ondemand before the upgrade.

**Comment by lukego** [comment]
Go ahead :).

**Comment by lukego** [comment]
If you want to account for performance variations you might want to run `perf stat` for visibility into the actual number of cycles used and the effective clock speed. This should be predictable with turbo disabled but visibility is always nice. (There is also a program called `i7z` that can monitor turbo behavior.)

One potential gotcha is that I believe the CPU frequency can still be throttled *downwards* when turbo is disabled but I would not expect this to happen during benchmark workloads.

**Comment by kbara** [comment]
I'll need to play with perf stat and i7z; it's been on my todo list for a while. :-)
```
# for CPUFREQ in /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor; do [ -f $CPUFREQ ] || continue; echo -n performance > $CPUFREQ; done 
```
makes a fairly dramatic difference, both in reported current CPU frequencies and in actual performance.

```
$ for x in {1..10}; do lwaftr-run; done | lwstats.py
Initial v4 MPPS: min: 1.462, max: 2.159, avg: 2.014, stdev: 0.2865 (n=10)
Initial v4 Gbps: min: 5.965, max: 8.809, avg: 8.216, stdev: 1.1688 (n=10)
Initial v6 MPPS: min: 1.462, max: 2.019, avg: 1.902, stdev: 0.2278 (n=10)
Initial v6 Gbps: min: 6.900, max: 9.531, avg: 8.980, stdev: 1.0754 (n=10)
  Final v4 MPPS: min: 1.486, max: 2.178, avg: 2.021, stdev: 0.2847 (n=10)
  Final v4 Gbps: min: 6.062, max: 8.885, avg: 8.245, stdev: 1.1615 (n=10)
  Final v6 MPPS: min: 1.486, max: 2.036, avg: 1.911, stdev: 0.2258 (n=10)
  Final v6 Gbps: min: 7.012, max: 9.610, avg: 9.019, stdev: 1.0659 (n=10)
```

**Comment by lukego** [comment]
My mental model is that packet processing performance tends to be linearly proportional to CPU frequency i.e. cycles/packet is fairly stable. This makes Mpps an awkward metric: it only makes sense with reference to a specific clock speed and that varies depending on CPU model and TurboBoost behavior.

In the NFV application we ultimately dealt with that by using a reference CPU that was the same as the one that Intel were using in their NFV case studies i.e. 3.0 GHz E5-2690v2. For development we used grindelwald that had slightly slower cores (2.7 GHz E5-2697v2) and so it worked out well: we hit the performance targets on the lab machine and so we had some extra safety margin on the reference machine.

So zooming waaay out it seems to me roughly like:

1. Software performance is best characterized by cycles/packet.
2. If you know what Mpps you need and you know the software's required cycles/packet budget then that tells you what GHz CPU you need to buy :-)

Of course in practice one often ends up testing with suboptimal hardware has to be creative.

**Comment by kbara** [comment]
I fully agree with you on all of that, but for various convoluted reasons, MPPS on a CPU identical to Interlaken's (with unspecified TurboBoost/CPU governor/etc behaviour) is what I currently have to deal with. :-)

**Pull request merged**